### PR TITLE
Soft delete subscriptions

### PR DESCRIPTION
**Pull request by gimenete**
Closes https://github.com/github/ecosystem-primitives/issues/103

A way of soft-deleting subscriptions could have been adding a `disabled` column or something like that. But I thought it was better and simpler to just create a new table storing that historic information for a couple of reasons:

- Security: we will never forget to specify `disabled: false` when querying subscriptions
- Performance: the table of deleted subscriptions will grow and grow, but there won't be a performance impact when querying active subscriptions.

I've decided to use the hooks feature from Sequelize so the implementation is as transparent and reusable as possible and I've added tests in all those cases I've found that subscriptions are deleted.

### CODE REVIEW

**Comment by tcbyrd** [comment]
Like the implementation here and using hooks. The way this is configured now, if a user re-subscribes, they would be starting over from scratch, but all their previous settings would be stored in the `DeletedSubscriptions` table under a different ID, correct? I wonder if we should a prompt for the user to restore settings from a previously deleted subscription?

**Comment by gimenete** [comment]
Hmmm restoring the settings would be a new feature and I'm not sure if it's worth it. Here the idea is just to keep a log of deleted subscriptions so if we need to investigate an incident we can match the deliveries with subscriptions, even if they were deleted at some point.

File: db/migrations/20181221164327-create-deleted-subscription.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.createTable('DeletedSubscriptions', {
+    id: {
+      allowNull: false,
+      autoIncrement: true,
+      primaryKey: true,
+      type: Sequelize.BIGINT,
+    },
```
Comment (wilhelmklopp) [inline db/migrations/20181221164327-create-deleted-subscription.js]: should we also have a column that represents the `id` from the `Subscriptions` table? Since the primary id in this new table will likely have a different value

File: test/integration/notifications.test.js [hunk 2]
```diff
@@ -494,7 +495,7 @@ describe('Integration: notifications', () => {
       });
     });
 
-    test('post prompt to re-subscribe after user loses access to repo', async () => {
+    test.only('post prompt to re-subscribe after user loses access to repo', async () => {
```
Comment (wilhelmklopp) [inline test/integration/notifications.test.js]: rogue `only` -- this is probably causing the drop in test coverage?

File: db/migrations/20181221164327-create-deleted-subscription.js (hunk 1 above)
Comment (gimenete) [reply db/migrations/20181221164327-create-deleted-subscription.js]: hmmm, what do you mean? I don't think we care about the original subscription id. But if we care I think the easiest solution would be to make the id the same and drop the `autoIncrement` in the `DeletedSubscriptions` table. What do you think?

File: test/integration/notifications.test.js (hunk 2 above)
Comment (gimenete) [reply test/integration/notifications.test.js]: Whooops. Thanks for spotting that!

File: db/migrations/20181221164327-create-deleted-subscription.js (hunk 1 above)
Comment (wilhelmklopp) [reply db/migrations/20181221164327-create-deleted-subscription.js]: > hmmm, what do you mean? I don't think we care about the original subscription id.

I was thinking from the perspective of needing to reconcile a deleted subscription with perhaps a log message somewhere that includes just the subscription id (but nothing else).

> But if we care I think the easiest solution would be to make the id the same and drop the autoIncrement in the DeletedSubscriptions table. What do you think?

Agree, this seems like the better solution 👍

**Comment by gimenete** [comment]
I'm not going to merge this yet until the PR about adding names to the db is merged, because I'll need to add those new columns in the deleted subscriptions too.

File: lib/activity/router.js [hunk 3]
```diff
@@ -57,13 +57,13 @@ module.exports = function route(callback) {
...
                 githubId: subscription.githubId,
+                githubName: subscription.githubName,
                 workspaceId: subscription.SlackWorkspace.slackId,
               },
             }, 'User lost access to resource. Deleting subscription.');
 
             await Promise.all([
-              // @todo: deactive this subscription instead of deleting the db record
-              await subscription.destroy(),
+              await subscription.destroyWithReason('lost access'),
```
Comment (wilhelmklopp) [inline lib/activity/router.js]: could we update this to `subscription creator lost access`?

```suggestion
              await subscription.destroyWithReason('subscription creator lost access'),
```

File: lib/slack/uninstall.js [hunk 4]
```diff
@@ -14,7 +14,7 @@ module.exports = async (req, res) => {
   });
 
   if (subscriptions) {
-    await Promise.all(subscriptions.map(subscription => subscription.destroy()));
+    await Promise.all(subscriptions.map(subscription => subscription.destroyWithReason('uninstall')));
```
Comment (wilhelmklopp) [inline lib/slack/uninstall.js]:
```suggestion
    await Promise.all(subscriptions.map(subscription => subscription.destroyWithReason('slack app uninstalled')));
```

File: db/migrations/20181221164327-create-deleted-subscription.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      allowNull: false,
+      type: Sequelize.DATE,
+    },
+    reason: {
+      type: Sequelize.STRING,
+      allowNull: true,
+    },
+  }),
+  down: queryInterface => queryInterface.dropTable('DeletedSubscriptions'),
+};
```
Comment (wilhelmklopp) [inline db/migrations/20181221164327-create-deleted-subscription.js]: ⚠️ scope creep alert ⚠️ 

Do we have any way of making sure that the columns stay in sync even when someone is making changes to the Subscriptions table?
Perhaps a test that asserts that the columns are the same (accounting for `reason`, `deleatedAt`, etc.)

File: db/migrations/20181221164327-create-deleted-subscription.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      type: Sequelize.JSON,
+    },
+    createdAt: {
+      allowNull: false,
+      type: Sequelize.DATE,
+    },
+    updatedAt: {
+      allowNull: false,
+      type: Sequelize.DATE,
+    },
```
Comment (wilhelmklopp) [inline db/migrations/20181221164327-create-deleted-subscription.js]: do we want a `deletedAt` column? (assuming that the values from `createdAt` and `updatedAt` are copied directly from the real subscription)

File: lib/models/subscription.js [hunk 7]
```diff
@@ -133,6 +139,17 @@ module.exports = (sequelize, DataTypes) => {
       this.changed('settings', true);
     },
 
+    async updateGithubName(githubName) {
+      if (this.githubName !== githubName) {
+        await this.update({ githubName });
+      }
+    },
```
Comment (wilhelmklopp) [inline lib/models/subscription.js]: 💡 I think it might be good to protect against accidentally updating to `null` here. if the passed in `githubName` is `null` (passed by accident from the calling function), then we should probably abort and log a warning?

File: lib/models/subscription.js [hunk 8]
```diff
@@ -173,7 +190,12 @@ module.exports = (sequelize, DataTypes) => {
     },
 
     async unsubscribe(githubId, channelId, slackWorkspaceId) {
-      await this.destroy({ where: { githubId, channelId, slackWorkspaceId } });
+      const where = { githubId, channelId, slackWorkspaceId };
+      const subscriptions = await this.findAll({ where });
+      await this.destroy({ where, individualHooks: true });
+
+      const ids = subscriptions.map(subscription => subscription.id);
+      await sequelize.model('DeletedSubscription').update({ reason: 'unsubscribe' }, { where: { id: ids } });
```
Comment (wilhelmklopp) [inline lib/models/subscription.js]: it'd be cool if we could pass a reason + multiple subscriptions into `destroyWithReason`, otherwise this could be a bit race-conditiony?

File: lib/models/subscription.js [hunk 9]
```diff
@@ -197,7 +219,10 @@ module.exports = (sequelize, DataTypes) => {
       });
       await this.destroy({
         where: { creatorId },
+        individualHooks: true,
       });
+      const ids = subscriptions.map(subscription => subscription.id);
+      await sequelize.model('DeletedSubscription').update({ reason: 'signout' }, { where: { id: ids } });
```
Comment (wilhelmklopp) [inline lib/models/subscription.js]: same question/concern here as above

File: db/migrations/20181221164327-create-deleted-subscription.js (hunk 5 above)
Comment (gimenete) [reply db/migrations/20181221164327-create-deleted-subscription.js]: There's at least one test that validates all the fields of the subscription.

https://github.com/integrations/slack/pull/754/files#diff-d662e6c04a8d262a012ac1efa9a6f9f7R103

The `reason` column is tested in many (all?) tests regarding deleted subscriptions.

I can be more explicit about other tests where I just test the count() of deletedsubscriptions, but since all the deleted subscriptions are created the same way (with `create(subscription.dataValues)` I didn't see it as necessary.

File: db/migrations/20181221164327-create-deleted-subscription.js (hunk 6 above)
Comment (gimenete) [reply db/migrations/20181221164327-create-deleted-subscription.js]: Oh, that's a good idea. I'll add that.

Comment (gimenete) [reply db/migrations/20181221164327-create-deleted-subscription.js]: Done :)

File: lib/models/subscription.js (hunk 7 above)
Comment (gimenete) [reply lib/models/subscription.js]: I added a conditional. I don't think we have to abort because if some webhook changes or a new webhook event has a different _shape_ we might be aborting the deletion for not a good reason.

File: lib/models/subscription.js (hunk 8 above)
Comment (gimenete) [reply lib/models/subscription.js]: I'm not doing that because I wanted to keep the bulk delete. I think this is very covered by tests so I would keep it like it is.
